docs: add 3.6 json templating upgrading notes (#13720)
argoproj/argo-workflows: docs/upgrading.md

## Upgrading to v3.6

> ### JSON templating fix

Reviewer: I'd probably move this near the legacy `patch` `pod` removal if the thought is that it shouldn't have much impact. I.e. order of importance (then all the metrics changes as things after it could be missed accidentally since it's a large section)

Author: I've moved it to the bottom as least important.

> When accessing a structure or array from json structure in a `jsonpath` expression you could end up getting a golang representation of the structure in error.

Reviewer: While the test cases in #12909 are `jsonpath` specific, the code is not, as far as I understand, otherwise it could have been applied to the `jsonpath` helper specifically.

The wording is also duplicative and inconsistent:
```suggestion
When accessing a map or array from JSON in a `jsonpath` expression, you would get a Golang representation.
```
- consistently use "JSON", not both "json" and "JSON" (esp in the same section)
  - "Golang" should similarly be capitalized as a proper noun
- simplify, per k8s style guide
  - "structure" is not needed (was previously used 3 times in one sentence) and is ambiguous, whereas "map" is a data structure consistent with "array"
  - "in error" is confusing/ambiguous as there is no error message. It's not needed either, we can just say what was returned before and what it returned now.

Author: Done, and slightly simplified again

> This will now return the json representation as hoped - see #12909.

Reviewer: ```suggestion
This now returns plain JSON.
```

- simplify, per k8s style guide

I'm also not sure we need to reference the PR and I'm not sure that referencing it would be helpful. They sometimes are and aren't referenced in the upgrading notes, but when they are it is part of the section heading and not in the contents. A header reference makes more sense to me as a way to correlate to the changelog items.

Also the PR number is not automatically linked

Author: Done
